Jigsaw mesh creation

sbrus89 opened the pull request:
This pull request adds the capabilities to do coastal refined meshes using the jigsaw workflow.

pwolfram commented:
@mark-petersen, I'm proposing we keep this PR over #59 from which it is derived (accounting for updates to recent ocean/develop).

pwolfram commented:
@mark-petersen, just a friendly follow-up reminder about getting your review and so that this can be merged in this week...

mark-petersen commented:
Thanks.  It is on my short list for today.

pwolfram commented:
Thanks @mark-petersen, we are looking forward to your feedback!

mark-petersen commented:
@sbrus89 could you test the head of this PR to generate your standard coastal-refined mesh?

mark-petersen commented:
@pwolfram I'm running some tests on this, will get it merged in.  Passed nightly regression suite.

Do you remember what `MpasMeshGenerator.sh` refers to here?  It is in ocean/develop and unchanged in the PR, but it doesn't seem to point to anything.  
```
testing_and_setup/compass/general.config.ocean:mesh_generator = FULL_PATH_TO_EXECUTABLE_MESH_GENERATION_SCRIPT
testing_and_setup/compass/ocean/soma/4km/32to4km/config_init1.xml:      <add_executable source="mesh_generator" dest="MpasMeshGenerator.sh"/>
testing_and_setup/compass/ocean/soma/8km/32to8km/config_init1.xml:      <add_executable source="mesh_generator" dest="MpasMeshGenerator.sh"/>
```
Did that point to the old jigsaw calling scripts?  If so, I can leave it.  I just want to remember to delete this line in a later commit, because right now the user doesn't know how to fill it in:
```
testing_and_setup/compass/general.config.ocean:mesh_generator = FULL_PATH_TO_EXECUTABLE_MESH_GENERATION_SCRIPT
```

mark-petersen commented:
@sbrus89 and @pwolfram Two questions:
1. I would like to add all new cases into a new directory, so that we have two directories:
-  everything with previous method will be left here:
`testing_and_setup/compass/ocean/global_ocean`
- everything created by jigsaw will be here: 
`testing_and_setup/compass/ocean/global_ocean_jigsaw`
That way, it is clear separation between the old and new process, and I can see which cases need to be 'converted' to the new process.  @sbrus89 I assume your *JS directories just mean jigsaw?

2. I finally came up with an alphabetized set of names for directories:
-  base_mesh
- culled_mesh
- initial_state
- prep_spinup1
- prep_spinup2
- simulation

where the directory formerly named `forward` is now `simulation`.  The `prep_` prefix is for order, but we can then use `prep_XXX` for anything else as well, before the big run.

I  will proceed with those (nice easy work at the conference), but I'm happy to hear feedback.

pwolfram commented:
@mark-petersen, along these lines, I already used jigsaw a few months ago for SOMA (https://github.com/MPAS-Dev/MPAS-Model/tree/master/testing_and_setup/compass/ocean/soma/4km/32to4km) but this is using the old way to setup test cases with jigsaw, so this gives a clear separation between what needs done in the long term to migrate to only using JIGSAW.

I think this organization is ok, but it may be a little artificial and may make things strange when building test suites in the short term.  There is a danger here too of having workflow divergence, e.g., supporting two ways of doing things may give people too many options.  I would propose all new test cases must use the new system.  

Also, once we have cleaned out `global_ocean`, I'd recommend moving `global_ocean_jigsaw` back to `global_ocean`.  Is this what you were thinking?

pwolfram commented:
@mark-petersen, `MpasMeshGenerator.sh` was from the previous workflow and pointed to https://github.com/MPAS-Dev/MPAS-Tools/blob/master/mesh_tools/triangle_jigsaw_to_netcdf/build_mesh.sh.   This was the previous way I built meshes with JIGSAW for the Q2 workflow.  I think we should delete this once we migrate the SOMA 32to4 and 32to8 cases over to the new system because it isn't in use anywhere else.

sbrus89 commented:
@mark-petersen Yes, the JS in the directory names is for jigsaw. I believe the only one that should have that is the EC60to30JS just to differentiate it from the original version. The other jigsaw meshes I added all begin with USDE or USNA since they are coastal refinements of either Delaware Bay or the US North American coast.

mark-petersen commented:
@pwolfram, I do want to migrate all global ocean cases to jigsaw, but need to have the old workflow for comparison, and to reproduce V1 meshes, at least for a while.  Having separate directories makes it more clear which process we are using.  It would be easy to change the name back later.

pwolfram commented on testing_and_setup/compass/ocean/global_ocean/EC60to30JS/init/define_base_mesh.py:
```diff
@@ -0,0 +1,10 @@
+def cellWidthVsLatLon():
+    params = mdt.params
+    lon,lat,cell_width = mdt.create_background_mesh(params["grd_box"],params["ddeg"],params["mesh_type"],params["dx_max"])
+    cell_width = cell_width/1000
```
May want to make this 1000.0 just to be clear it is a float (and to avoid integer division issues).

pwolfram commented on testing_and_setup/compass/ocean/global_ocean/QU15/init/config_base_mesh.xml:
```diff
@@ -0,0 +1,21 @@
+	<add_link source_path="mpas_model" source="testing_and_setup/compass/ocean/global_ocean/QU15/init/define_base_mesh.py" dest="define_base_mesh.py"/>
+	<add_link source_path="mpas_tools" source="visualization/paraview_vtk_field_extractor/paraview_vtk_field_extractor.py" dest="paraview_vtk_field_extractor.py"/>
+	<add_link source_path="jigsaw-geo-matlab" source="." dest="jigsaw-geo-matlab"/>
+	<add_link source_path="jigsaw-geo-matlab" source="jigsaw/geo/topo.msh" dest="topo.msh"/>
```
Should the lines above be templated to have a general header?

pwolfram commented on testing_and_setup/compass/ocean/global_ocean/QU240/init/config_base_mesh.xml:
```diff
@@ -0,0 +1,21 @@
+		<step executable="./build_mesh.py">
+		</step>
+	</run_script>
+</config>
```
It seems like this entire script is a complete copy.  Perhaps this is ok but it may be worth only having one master copy of this type of code around for clarity.

mark-petersen replied:
Done.

pwolfram commented on testing_and_setup/compass/ocean/jigsaw_to_MPAS/build_mesh.py:
```diff
@@ -0,0 +1,89 @@
+print "running", ' '.join(args)
+subprocess.check_call(args, env=os.environ.copy())
+
+#print 'Step 6. Create vtk file for visualization'
```
Should this be made as a command line argument option?  Right now it appears to be commented out for debugging purposes.

pwolfram replied:
Ok for now

pwolfram commented on testing_and_setup/compass/ocean/jigsaw_to_MPAS/build_mesh_compass.py:
```diff
@@ -0,0 +1,59 @@
+print "running", ' '.join(args)
+subprocess.check_call(args, env=os.environ.copy())
+
+#print 'Step 5. Create vtk file for visualization'
```
Same comment as previous-- should this be a command line argument option?

pwolfram commented on testing_and_setup/compass/ocean/jigsaw_to_MPAS/mesh_definition_tools.py:
```diff
@@ -0,0 +1,903 @@
+#!/usr/bin/env python
+'''
+name: mesh_definition_tools
+authors: Mark R. Petersen
```
@sbrus89, please add your name here.

pwolfram commented on testing_and_setup/compass/ocean/jigsaw_to_MPAS/mesh_definition_tools.py:
```diff
@@ -0,0 +1,903 @@
+  cellWidthOut = cellWidthPole/np.power(densityRRS,0.25)
+  return cellWidthOut  
+
+#'''
```
Do we want all this commented out code to be included in this PR?

pwolfram commented on testing_and_setup/compass/ocean/jigsaw_to_MPAS/mesh_definition_tools.py:
```diff
@@ -0,0 +1,903 @@
+#  plot_coarse_coast(plot_box)
+#  plt.colorbar()
+#  plt.axis('equal')
+#  plt.savefig('bckgnd_grid_bathy_grad.png',bbox_inches='tight')
```
Should this commented out code above be included in the PR?

pwolfram commented:
@mark-petersen and @sbrus89, I had a few minor comments that may be worth considering prior to a final merge.  Thanks for all your hard work on this process!

mark-petersen commented:
Also, if you look in the QU60 and QU240 directories, I really simplified things, and made links to standard config files where I could.  @sbrus89 I was afraid to change your US* files any more because I didn't want to break them.  But you might find that you can simplify some of your files by just linking in as well.

sbrus89 commented on testing_and_setup/compass/ocean/jigsaw_to_MPAS/coastal_tools.py:
```diff
@@ -0,0 +1,598 @@
+params = {
+
+# Path to bathymetry data and name of file
+"nc_file": "/users/sbrus/climate/bathy_data/SRTM15_plus/earth_relief_15s.nc",
```
I agree. I'll change the name to default_params, since each case's define_base_mesh.py imports and then modifies each option.

Can I add the bathy data file to the initial_condition_database so it can be linked with the config_base_mesh.xml file?

sbrus89 commented on testing_and_setup/compass/ocean/global_ocean/USDEQU120cr10rr1/init/define_base_mesh.py:
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env python
+
+import numpy as np
+import coastal_tools as ct
```
Thanks!

sbrus89 commented:
@mark-petersen I think I addressed most of the issues that you and Phil pointed out. I tested all three of my coastal mesh cases and they are working as expected. Let me know if there's anything else I can do on my end to help! Thanks again.

pwolfram commented:
@mark-petersen, are we ready to merge this now?  I can get in some of my other test cases from here.  Thanks!

sbrus89 commented:
@pwolfram I ran a final check and everything works as expected. The autopep8 changes made some strange formatting choices in some places in coastal_tools.py. I have a commit that cleans some of that up, but I don't have to include it if you'd rather leave things as-is.

pwolfram commented:
@sbrus89, please include the commit and I'll test once you push it.

sbrus89 commented:
@pwolfram, Sounds good. I just did.

pwolfram commented:
Awesome, we sorted through the key issues with a fresh checkout on IC:

![image](https://user-images.githubusercontent.com/4295853/46048918-047cc980-c0e9-11e8-9b11-2fdf193ef50b.png)
